### ISSUE TITLE
Rewrite extension scripts in TypeScript

### ISSUE THREAD

**Issue by unflxw** [appsignal/appsignal-nodejs#598]
The extension install script is written in JavaScript, while the rest of our codebase is in TypeScript. If we could move this to TypeScript as well, it would streamline using bits of the codebase, such as the Transmitter, from the extension script.

We may want to reconsider whether the separation between the scripts and the regular source code is necessary. If we keep this separation, this may mean our build process calls the TypeScript compiler twice. There may be some TSConfig toggle that allows us to work around this.

**Comment by tombruijn** [comment]
If we are able to configure TypeScript to compile everything with one config, the one thing we have to watch out for is that users can't actually import or run the extension install script files. They should never be accessible outside of the package.

**Comment by unflxw** [comment]
> the one thing we have to watch out for is that users can't actually import or run the extension install script files.

While looking into an issue regarding making the extension use the transmitter (which relates to it, as it's in TS), I realised that the above is already the case in our current releases:

```js
const reportScript = require("@appsignal/nodejs/scripts/extension/report")
console.log("reportPath", reportScript.reportPath())
```

**Comment by tombruijn** [comment]
Oh that's maybe not the best setup. We can move the function to the `src` dir and import in the install script instead? Turn it around.

**Comment by tombruijn** [comment]
As just discussed in the call, this would not work for the local setup. Not from scratch. If you clone the repo, run `mono bootstrap`, then it will run `npm run install` on the `@appsignal/nodejs` package, which won't have the TypeScript compiled files yet.

**Comment by unflxw** [comment]
That's true, but we could have `npm install` run `npm run build` when it's a local build. [The branch where the extension uses the transmitter does this to work around it.](https://github.com/appsignal/appsignal-nodejs/commit/438f4f3a0eed555fd5bf9718a7f2d08cc2bc5f8e)

I think the above issue, with the scripts being importable, is essentially inevitable. I've checked `node-sass`, which also has scripts, and the same thing happens. `require()` allows you to access any file in the modules, and I don't know if there's a way around that. That said, [the changes to the extension](https://github.com/appsignal/appsignal-nodejs/compare/test-extension-download#diff-40660fa004b190328001c758a42187d78b465ec9330f6dd4f9d8412235f36c17R276) in #600 should at least make it so that requiring it doesn't automatically run the install process.

**Comment by tombruijn** [comment]
Let's postpone work on it for a bit until we've completed the work on the extension install script and are a bit more sure what we want here.

**Comment by tombruijn** [comment]
Noemi and I discussed fixing the `mono bootstrap` flow that would make this whole thing less confusing.

`mono bootstrap` currently calls `npm run install` which installs the extension. If we use TypeScript in the extension install script you need to build the TypeScript to JavaScript first.

What if instead we don't run `npm run install` on `mono bootstrap` locally? But run `npm run install` after `npm run build` using mono (using a post build hook) or package.json (`postbuild: "npm run install"`). That way it's not automatically installed on `mono bootstrap` and we don't have this build order issue locally.

This would only be for local builds, because the distributed package does need to have this flow, but it ships with the TypeScript compiled to JavaScript.
